File: ios/HivePulse/ViewModels/AdminViewModel.swift
import Foundation

// MARK: - Stats

@MainActor
final class AdminStatsViewModel: ObservableObject {
    @Published var stats: PlatformStats?
    @Published var tokenStats: AdminTokenStats?
    @Published var isLoading = false
    @Published var errorMessage: String?
    @Published var preset = "30d"

    private let service: any AdminServiceProtocol

    init(service: any AdminServiceProtocol = AdminService()) {
        self.service = service
    }

    func load() async {
        isLoading = true
        errorMessage = nil
        do {
            async let s = service.getStats(preset: preset)
            async let t = service.getTokenStats()
            (stats, tokenStats) = try await (s, t)
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }
}

// MARK: - Users

@MainActor
final class AdminUsersViewModel: ObservableObject {
    @Published var users: [AdminUserOut] = []
    @Published var isLoading = false
    @Published var errorMessage: String?
    @Published var search = ""
    @Published var filterSupporter: Bool? = nil
    @Published var page = 1
    @Published var pages = 1

    private let service: any AdminServiceProtocol

    init(service: any AdminServiceProtocol = AdminService()) {
        self.service = service
    }

    func load() async {
        isLoading = true
        errorMessage = nil
        do {
            let resp = try await service.getUsers(page: page, search: search, isSupporter: filterSupporter)
            users = resp.items
            pages = resp.pages
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }

    func toggleSupporter(_ user: AdminUserOut) async {
        do {
            let updated = try await service.setSupporter(userId: user.id, isSupporter: !user.isSupporter)
            if let idx = users.firstIndex(where: { $0.id == user.id }) {
                users[idx] = updated
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func revokeTokens(_ userId: String) async {
        do {
            try await service.revokeTokens(userId: userId)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func deleteUser(_ userId: String) async {
        do {
            try await service.deleteUser(userId: userId)
            users.removeAll { $0.id == userId }
        } catch {
            errorMessage = error.localizedDescription
        }
    }
}

// MARK: - Map

@MainActor
final class AdminMapViewModel: ObservableObject {
    @Published var apiaries: [AdminApiary] = []
    @Published var isLoading = false
    @Published var errorMessage: String?
    @Published var showFlagged = false
    @Published var page = 1
    @Published var pages = 1

    private let service: any AdminServiceProtocol

    init(service: any AdminServiceProtocol = AdminService()) {
        self.service = service
    }

    func load() async {
        isLoading = true
        errorMessage = nil
        do {
            let resp = showFlagged
                ? try await service.getFlaggedApiaries(page: page)
                : try await service.getApiaries(page: page)
            apiaries = resp.items
            pages = resp.pages
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }

    func setPrivate(_ apiary: AdminApiary) async {
        do {
            try await service.setPrivate(apiaryId: apiary.id)
            apiaries.removeAll { $0.id == apiary.id }
        } catch {
            errorMessage = error.localizedDescription
        }
    }
}

// MARK: - Health

@MainActor
final class AdminHealthViewModel: ObservableObject {
    @Published var summary: HealthSummary?
    @Published var inactiveUsers: [InactiveUser] = []
    @Published var noVarroaApiaries: [NoVarroaApiary] = []
    @Published var zeroInspectionHives: [ZeroInspectionHive] = []
    @Published var isLoading = false
    @Published var isDrillLoading = false
    @Published var errorMessage: String?
    @Published var activeDetail: HealthDetail? = nil
    @Published var inactivePage = 1
    @Published var inactivePages = 1

    enum HealthDetail { case inactive, noVarroa, zeroHives }

    private let service: any AdminServiceProtocol

    init(service: any AdminServiceProtocol = AdminService()) {
        self.service = service
    }

    func load() async {
        isLoading = true
        errorMessage = nil
        do {
            summary = try await service.getHealthSummary()
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }

    func toggleDetail(_ detail: HealthDetail) async {
        if activeDetail == detail {
            activeDetail = nil
            return
        }
        activeDetail = detail
        isDrillLoading = true
        defer { isDrillLoading = false }
        do {
            switch detail {
            case .inactive:
                let resp = try await service.getInactiveUsers(page: inactivePage)
                inactiveUsers = resp.items
                inactivePages = resp.pages
            case .noVarroa:
                noVarroaApiaries = try await service.getNoVarroaApiaries()
            case .zeroHives:
                let resp = try await service.getZeroInspectionHives(page: 1)
                zeroInspectionHives = resp.items
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func loadMoreInactive() async {
        guard inactivePage < inactivePages else { return }
        inactivePage += 1
        isDrillLoading = true
        do {
            let resp = try await service.getInactiveUsers(page: inactivePage)
            inactiveUsers += resp.items
            inactivePages = resp.pages
        } catch {
            errorMessage = error.localizedDescription
        }
        isDrillLoading = false
    }
}

File: android/app/src/main/java/com/hivepulse/app/ui/admin/AdminViewModel.kt
package com.hivepulse.app.ui.admin

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hivepulse.app.data.api.*
import com.hivepulse.app.data.repository.AdminRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import javax.inject.Inject

// MARK: - Stats

data class AdminStatsState(
    val stats: PlatformStats? = null,
    val tokenStats: AdminTokenStats? = null,
    val isLoading: Boolean = false,
    val error: String? = null,
    val preset: String = "30d"
)

@HiltViewModel
class AdminStatsViewModel @Inject constructor(
    private val repo: AdminRepository
) : ViewModel() {
    private val _state = MutableStateFlow(AdminStatsState())
    val state = _state.asStateFlow()

    init { load() }

    fun load() = viewModelScope.launch {
        _state.update { it.copy(isLoading = true, error = null) }
        runCatching {
            val s = repo.getStats(state.value.preset)
            val t = repo.getTokenStats()
            Pair(s, t)
        }.onSuccess { (s, t) ->
            _state.update { it.copy(isLoading = false, stats = s, tokenStats = t) }
        }.onFailure { e ->
            _state.update { it.copy(isLoading = false, error = e.message) }
        }
    }

    fun setPreset(preset: String) {
        _state.update { it.copy(preset = preset) }
        load()
    }

    fun clearError() = _state.update { it.copy(error = null) }
}

// MARK: - Users

data class AdminUsersState(
    val users: List<AdminUserOut> = emptyList(),
    val isLoading: Boolean = false,
    val error: String? = null,
    val search: String = "",
    val filterSupporter: Boolean? = null,
    val page: Int = 1,
    val pages: Int = 1
)

@HiltViewModel
class AdminUsersViewModel @Inject constructor(
    private val repo: AdminRepository
) : ViewModel() {
    private val _state = MutableStateFlow(AdminUsersState())
    val state = _state.asStateFlow()

    init { load() }

    fun load() = viewModelScope.launch {
        _state.update { it.copy(isLoading = true, error = null) }
        runCatching {
            repo.getUsers(state.value.page, state.value.search, state.value.filterSupporter)
        }.onSuccess { resp ->
            _state.update { it.copy(isLoading = false, users = resp.items, pages = resp.pages) }
        }.onFailure { e ->
            _state.update { it.copy(isLoading = false, error = e.message) }
        }
    }

    fun search(query: String) {
        _state.update { it.copy(search = query, page = 1) }
        load()
    }

    fun prevPage() {
        if (state.value.page <= 1) return
        _state.update { it.copy(page = it.page - 1) }
        load()
    }

    fun nextPage() {
        if (state.value.page >= state.value.pages) return
        _state.update { it.copy(page = it.page + 1) }
        load()
    }

    fun toggleSupporter(user: AdminUserOut) = viewModelScope.launch {
        runCatching { repo.setSupporter(user.id, !user.isSupporter) }
            .onSuccess { updated ->
                _state.update { s -> s.copy(users = s.users.map { if (it.id == updated.id) updated else it }) }
            }
            .onFailure { e -> _state.update { it.copy(error = e.message) } }
    }

    fun revokeTokens(userId: String) = viewModelScope.launch {
        runCatching { repo.revokeTokens(userId) }
            .onFailure { e -> _state.update { it.copy(error = e.message) } }
    }

    fun deleteUser(userId: String) = viewModelScope.launch {
        runCatching { repo.deleteUser(userId) }
            .onSuccess { _state.update { it.copy(users = it.users.filter { u -> u.id != userId }) } }
            .onFailure { e -> _state.update { it.copy(error = e.message) } }
    }

    fun clearError() = _state.update { it.copy(error = null) }
}

// MARK: - Map

data class AdminMapState(
    val apiaries: List<AdminApiary> = emptyList(),
    val isLoading: Boolean = false,
    val error: String? = null,
    val showFlagged: Boolean = false,
    val page: Int = 1,
    val pages: Int = 1
)

@HiltViewModel
class AdminMapViewModel @Inject constructor(
    private val repo: AdminRepository
) : ViewModel() {
    private val _state = MutableStateFlow(AdminMapState())
    val state = _state.asStateFlow()

    init { load() }

    fun load() = viewModelScope.launch {
        _state.update { it.copy(isLoading = true, error = null) }
        runCatching {
            if (state.value.showFlagged) repo.getFlaggedApiaries(state.value.page)
            else repo.getApiaries(state.value.page)
        }.onSuccess { resp ->
            _state.update { it.copy(isLoading = false, apiaries = resp.items, pages = resp.pages) }
        }.onFailure { e ->
            _state.update { it.copy(isLoading = false, error = e.message) }
        }
    }

    fun toggleFlagged(flagged: Boolean) {
        _state.update { it.copy(showFlagged = flagged, page = 1) }
        load()
    }

    fun prevPage() {
        if (state.value.page <= 1) return
        _state.update { it.copy(page = it.page - 1) }
        load()
    }

    fun nextPage() {
        if (state.value.page >= state.value.pages) return
        _state.update { it.copy(page = it.page + 1) }
        load()
    }

    fun setPrivate(apiary: AdminApiary) = viewModelScope.launch {
        runCatching { repo.setPrivate(apiary.id) }
            .onSuccess { _state.update { it.copy(apiaries = it.apiaries.filter { a -> a.id != apiary.id }) } }
            .onFailure { e -> _state.update { it.copy(error = e.message) } }
    }

    fun clearError() = _state.update { it.copy(error = null) }
}

// MARK: - Health

enum class HealthDetail { INACTIVE, NO_VARROA, ZERO_HIVES }

data class AdminHealthState(
    val summary: HealthSummary? = null,
    val inactiveUsers: List<InactiveUser> = emptyList(),
    val noVarroaApiaries: List<NoVarroaApiary> = emptyList(),
    val zeroInspectionHives: List<ZeroInspectionHive> = emptyList(),
    val isLoading: Boolean = false,
    val isDrillLoading: Boolean = false,
    val error: String? = null,
    val activeDetail: HealthDetail? = null,
    val inactivePage: Int = 1,
    val inactivePages: Int = 1
)

@HiltViewModel
class AdminHealthViewModel @Inject constructor(
    private val repo: AdminRepository
) : ViewModel() {
    private val _state = MutableStateFlow(AdminHealthState())
    val state = _state.asStateFlow()

    init { load() }

    fun load() = viewModelScope.launch {
        _state.update { it.copy(isLoading = true, error = null) }
        runCatching { repo.getHealthSummary() }
            .onSuccess { s -> _state.update { it.copy(isLoading = false, summary = s) } }
            .onFailure { e -> _state.update { it.copy(isLoading = false, error = e.message) } }
    }

    fun toggleDetail(detail: HealthDetail) = viewModelScope.launch {
        if (state.value.activeDetail == detail) {
            _state.update { it.copy(activeDetail = null) }
            return@launch
        }
        _state.update { it.copy(activeDetail = detail, isDrillLoading = true) }
        runCatching {
            when (detail) {
                HealthDetail.INACTIVE -> {
                    val resp = repo.getInactiveUsers(1)
                    _state.update { it.copy(inactiveUsers = resp.items, inactivePage = 1, inactivePages = resp.pages) }
                }
                HealthDetail.NO_VARROA -> {
                    val list = repo.getNoVarroaApiaries()
                    _state.update { it.copy(noVarroaApiaries = list) }
                }
                HealthDetail.ZERO_HIVES -> {
                    val resp = repo.getZeroInspectionHives(1)
                    _state.update { it.copy(zeroInspectionHives = resp.items) }
                }
            }
        }.onFailure { e -> _state.update { it.copy(error = e.message) } }
        _state.update { it.copy(isDrillLoading = false) }
    }

    fun loadMoreInactive() = viewModelScope.launch {
        val s = state.value
        if (s.inactivePage >= s.inactivePages) return@launch
        _state.update { it.copy(isDrillLoading = true) }
        runCatching { repo.getInactiveUsers(s.inactivePage + 1) }
            .onSuccess { resp ->
                _state.update {
                    it.copy(
                        inactiveUsers = it.inactiveUsers + resp.items,
                        inactivePage = resp.page,
                        inactivePages = resp.pages
                    )
                }
            }
            .onFailure { e -> _state.update { it.copy(error = e.message) } }
        _state.update { it.copy(isDrillLoading = false) }
    }

    fun clearError() = _state.update { it.copy(error = null) }
}
